### PR TITLE
Feature/ansible inventory

### PR DESCRIPTION
**Pull request by Donien**
### Features

- Use of cache via [Ansible Cache Plugins](https://docs.ansible.com/ansible/latest/collections/index_cache.html) to reduce frequency of API calls (`jsonfile` was tested)
- Filtering of hosts right when making the API call
  Filter by
  - host name
  - zone
  - group membership
  - custom (write the filter yourself)
  - custom variables
- Builds Ansible groups based on Icinga groups
- Use of [Ansible Constructed Plugin](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/constructed_inventory.html) to build groups based on available host variables (`keyed_groups`)
- Use of either host name, `address` or `address6` as `ansible_host`
- Use of arbitrary

Quick example of an inventory definition:  
**inventory-icinga.yml**

```yaml
---
plugin: icinga.icinga.icinga
url: https://example.com
user: root
password: root
validate_certs: true

# Create groups 'icinga_linux_distribution_debian', 'icinga_linux_distribution_rhel', etc.
keyed_groups:
  - prefix: "icinga_linux_distriubtion"
    key: vars.distribution

# Set 'ansible_user' equal to the host's custom variable 'vars.ansible_user'
ansible_user_var: "vars.ansible_user"

filters:
  # Only match hosts that are in zone 'satellite1'
  zone:
    - satellite1
  vars:
    match:
      # ... and run those distributions except 'ubuntu'
      linux_distribution:
        - '!ubuntu'
        - 'debian'
        - 'centos'
        - 'rhel'
        - 'suse'
    in:
      # ... and have on of those services in 'vars.services' array
      services:
        - 'dns'    
        - 'database'
    is:
      # ... and have variable 'vars.ansible_managed' set to true AND have variable 'ansible_user' set
      ansible_managed: true
      ansible_user: set
```

---

I believe this to be a quite useful feature that many people could benefit from.

If any changes would need to be made to the structure, let me know and I will gladly change things around to fit the needs.

### CODE REVIEW

File: .github/workflows/python-test.yml [hunk 1]
```diff
@@ -57,8 +57,13 @@ jobs:
         with:
           python-version: ${{ matrix.python }}
 
+      - name: Install dependencies
+        run: |
+          python -m pip install "ansible-core==2.14.7"
```
Comment (martialblog) [inline .github/workflows/python-test.yml:62]: Will this include the requests library?

File: README.md [hunk 2]
```diff
@@ -16,6 +16,7 @@ Collection to setup and manage components of the Icinga software stack.
 * [Role: icinga.icinga.icingaweb2](doc/role-icingaweb2/role-icingaweb2.md)
 * [Role: icinga.icinga.monitoring_plugins](doc/role-monitoring_plugins/role-monitoring_plugins.md)
   * [List of Available Check Commands](doc/role-monitoring_plugins/check_command_list.md)
+* [Inventory Plugin: icinga.icinga.icinga](doc/plugins/inventory/icinga-inventory-plugin.md)
```
Comment (martialblog) [inline README.md:19]: The docs should maybe mention that the requests library needs to be installed.

File: plugins/inventory/icinga.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+        sub_filter_positive = list()
+        sub_filter_negative = list()
+
+        for value in values:
+            tmp_string = filter_string_base.format(value, key)
+            if value.startswith('!'):
+                tmp_string = '!{}'.format(tmp_string.replace('!', ''))
+                sub_filter_negative.append(tmp_string)
+            else:
+                sub_filter_positive.append(tmp_string)
```
Comment (martialblog) [inline plugins/inventory/icinga.py:257]: Couldn't this be done as the default without the else?

File: plugins/inventory/icinga.py [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+            sub_filters.append(sub_filter_string)
+
+        if sub_filters:
+            filter_string = '({})'.format(')&&('.join(sub_filters))
+
+        return filter_string
+
+
+    def _get_hosts(self):
+        s   = self._get_session()
```
Comment (martialblog) [inline plugins/inventory/icinga.py:390]: Extra space

File: plugins/inventory/icinga.py [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+        except RequestException:
+            self.display.vvv('Error accessing \'{self.api_url}\'.')
+            raise
+
+        # Create filter
+        filter_string = self._create_filter()
+        if filter_string:
+            data = '{ "filter": "' + filter_string + '" }'
+        else:
+            data = None
```
Comment (martialblog) [inline plugins/inventory/icinga.py:407]: Could be the default without the else

**Comment by martialblog** [comment]
I think it's a very nice addition. I will give the code a closer look next week.

@Donien Just FYI, the coverage currently:

```
Name                                  Stmts   Miss  Cover   Missing
-------------------------------------------------------------------
plugins/inventory/icinga.py             264    149    44%   155-161, 189-190, 199-206, 280, 317-386, 390-417, 422-477, 487-488, 494-546
```

File: plugins/inventory/icinga.py (hunk 3 above)
Comment (Donien) [reply plugins/inventory/icinga.py:257]: Here the string is appended to different lists depending on whether the string starts with a '!' indicating negation. The negated and non-negated filters are concatenated differently.  
We could of course use a `continue` within the `if` block and remove the `else` if that's considered more readable.

**Review by martialblog — APPROVED** [review]
:+1: LGTM

**Comment by martialblog** [comment]
@mkayontour I think this is good to go.

**Comment by MarcusCaepio** [comment]
Just a question: There is already an icinga2 inventory plugin in the ansible community repo: https://docs.ansible.com/ansible/latest/collections/community/general/icinga2_inventory.html 
Why an additional one? :)

**Comment by Donien** [comment]
@MarcusCaepio @martialblog 

First of all: Yes, there already is a plugin that achieves basically the same goal.  

The 2 main differences are:  

- The way you call the plugin (especially looking at the filters)
- The outputs the plugins produce

In the plugin from this PR, Icinga 2 API filters are abstracted for certain criteria ( e.g. `zone: [ "main", "satellite1", "satellite2", ... ]` ), (IMHO) making it easier to filter results without having to know the exact syntax the API needs.  

The second aspect are the host variables. The [community plugin](https://docs.ansible.com/ansible/latest/collections/community/general/icinga2_inventory.html) creates the following host variables (underneath a common key):

- address
- address6
- display_name
- groups
- name
- state
- state_type
- templates
- vars
- zone

The plugin from this PR takes all attributes of any given Icinga 2 host and puts them into host variables (with a common prefix).  
Some examples:  

- check_command
- command_endpoint
- last_check_result
- notes_url

---

I personally believe that an Ansible Collection should include everything that is relevant to the topic of that collection.  
In the case of the Icinga Collection, an inventory plugin fits neatly.  

I also think that we (people that work on this collection) should consider `ansible-core` as the starting point. So if people install this collection, they should have access to an inventory plugin for Icinga without having to install `ansible` with its collections.


Of course this is just my opinion and I'm open to discussion.

**Comment by MarcusCaepio** [comment]
What the community plugin serves and what not are in the user's hand and can be configured in the x.icinga2.yml config file. e.g. with the compose parameter and / or groups :) See the docs: https://docs.ansible.com/ansible/latest/collections/community/general/icinga2_inventory.html 

```
# my.icinga2.yml
plugin: community.general.icinga2
url: http://localhost:5665
user: ansible
password: secure
host_filter: \"linux-servers\" in host.groups
validate_certs: false
inventory_attr: name
groups:
  # simple name matching
  webservers: inventory_hostname.startswith('web')

  # using icinga2 template
  databaseservers: "'db-template' in (icinga2_attributes.templates|list)"

compose:
  # set all icinga2 attributes to a host variable 'icinga2_attrs'
  icinga2_attrs: icinga2_attributes

  # set 'ansible_user' and 'ansible_port' from icinga2 host vars
  ansible_user: icinga2_attributes.vars.ansible_user
  ansible_port: icinga2_attributes.vars.ansible_port | default(22)
```
A result with the compose part enabled, for example, looks like this:

```
{
    "ansible_host": "1.2.3.4",
    "display_name": "myserver.domain.com",
    "hostname": "asdf.domain.com",
    "icinga2_attrs": {
        "address": "1.2.3.4",
        "address6": "",
        "display_name": "myserver.domain.com",
        "groups": [],
        "name": "asdf.domain.com",
        "state": "on",
        "state_type": 1,
        "templates": [
            "asdf.domain.com",
            "default-host"
        ],
        "vars": {
            "agent_endpoint": "asdf.domain.com",
            "icmp_cpl": 100,
            "icmp_crta": 500,
            "icmp_packets": 1,
            "icmp_wpl": 100,
            "icmp_wrta": 200,
            "tcp": {
                "22": {
                    "check_interval": 3600,
                    "notes": "ssh",
                    "tcp_port": "22"
            }
        },
        "zone": "master"
    },
    "state": "on",
    "state_type": 1
}
```

Don't get me wrong, I have nothing against your pull request. I just think that if there is already a plugin that is in and maintained by the ansible community, then one should get together and extend that plugin if necessary (e.g. add the check_command to the output), rather than reinventing the wheel entirely. Or there should be a discussion between the ansible community and icinga, which plugin should be the one and be placed in the icinga collection

**Pull request merged**